README.md
# blitter

[![Current Crates.io Version](https://img.shields.io/crates/v/blitter.svg)](https://crates.io/crates/blitter)
[![Downloads badge](https://img.shields.io/crates/d/blitter.svg)](https://crates.io/crates/blitter)

This library performs various blitting and drawing operations on a raw 32 bits framebuffer, whatever the encoding.
- Bitmap blitting / cropping
- Blit a part of bitmap (ie. bitmap fonts)
- Blit with a color or bits mask
- Pixel plotting
- Optional PNG decoding feature

Example:
```
// Framebuffer initialization
let mut pixels: Vec<u32> = vec!(0; WIDTH * HEIGHT);
let mut fb = Framebuffer {width: WIDTH, height: HEIGHT, pixels: &mut pixels};

// For example, you can push all the bitmaps in a single vec to give ownership of all bitmaps
let mut bitmaps = Vec::new();
bitmaps.push(Bitmap {w: 10, h: 10, x: 0, y: 0, pixels: &image::PIXELS});

while *display loop with some display library* {
    blitter_test(&mut fb, &mut bitmaps);
    *your display lib display update function with buffer &fb.pixels*
}

// For testing : moves a 10x10 square and prints a 4x4 pixel at the center of the screen
fn blitter_test(mut fb: &mut Framebuffer, bitmaps: &mut Vec<Bitmap>) {
    fb.clear_area(640, 10, 0, 0, 0).unwrap();
    bitmaps[0].blit(&mut fb).unwrap();   //copies a bitmap to the framebuffer
    if bitmaps[0].x < WIDTH - 10 { bitmaps[0].x = bitmaps[0].x+3; } else { fb.clear(0); }
    fb.draw_fatpixel(320,240,4,0xffffffff).unwrap();
}
```

You can also view and run some (very basic) examples using the [minifb library](https://crates.io/crates/minifb) in the 'examples' directory:
```
cargo run --example demo --features="png-decode"
cargo run --example minifb --features="png-decode"
cargo run --example square
```

[![Screenshot](resources/screenshot.png)]

License: GPL-3.0


CHANGELOG.md
# Changelog

### v0.6.0 (2020-04-10)

- [Added] Support for RGBA PNG decoding
- [Added] Library example and screenshot

### v0.5.0 (2020-03-30)

- [Added] Automatic clipping of blitted bitmaps
- [Added] square & minifb mask examples
- [Changed] Bitmap coordinates can be negative
- [Changed] Blitting with mask + added mask enum
- [TODO] Bits mask import
- [TODO] More sexy examples

### v0.4.0 (2020-01-30)

- PNG decoding as 0RGB or RGBA

### v0.3.0 (2020-01-29)

- [Added] PNG decoding optional feature (WIP)
- [Fixed] Blitting functions
- [Fixed] Out of bounds checks

### v0.2.0 (2020-01-27)

- [Added] Blit with bit mask
- [Added] Blit with color mask
- [Added] Out of limits operations check
- [Added] minifb example

### v0.1.0 (2020-01-26)

- Initial release, bitmap blitting, framebuffer clear & partial clear, pixel drawing


examples/demo.rs
//#![windows_subsystem = "windows"]
use blitter::*;
use minifb::{Key, Scale, Window, WindowOptions};

const WIDTH: usize = 320;
const HEIGHT: usize = 256;
const TEXT: &'static str = "Blit,Blit with color mask,Automatic cropping,Blit with bits mask,Blit of a partial bitmap";

fn main() {
    // Framebuffer initialization
    let mut pixels: Vec<u32> = vec![0; WIDTH * HEIGHT];
    let mut fb = Framebuffer {
        width: WIDTH,
        height: HEIGHT,
        pixels: &mut pixels,
    };

    // Font generation
    let f = include_bytes!("../resources/FONT2_8X8.BIN");
    // To store the 96 * 8 * 8 pixels as 32 bits pixel data
    let mut font = vec![0; 6144];
    // Pixel counter for destination bitmap
    let mut c = 0;
    for byte in 0..8 {
        // index of first char byte
        for char_index in 0..96 {
            // for each bit
            for bit in 0..8 {
                if f[char_index * 8 + byte] << bit & 0x80u8 == 0x80u8 {
                    font[c] = 0xffffffffu32;
                } else {
                    font[c] = 0u32;
                }
                c += 1;
            }
        }
    }

    // Bitmap generation
    let path = "resources/littledragonG.png";
    let png = from_png_file(&path, PixelFormat::Zrgb).unwrap();
    // Bitmaps structs stored in a Vec, could be a hashmap, to give more easily ownership to other functions
    let mut bitmaps = Vec::new();
    bitmaps.push(Bitmap {
        w: 768,
        h: 8,
        x: 0,
        y: 0,
        pixels: &font,
    }); //bitmaps[0] is the font
    bitmaps.push(Bitmap{w: png.0, h: png.1, x: 0, y: 0, pixels: &png.2});

    let mut window = Window::new(
        "Blitter demo",
        WIDTH,
        HEIGHT,
        WindowOptions {
            scale: Scale::X2,
            ..WindowOptions::default()
        },
    )
    .unwrap_or_else(|e| {
        panic!("{}", e);
    });

    // Limit to max ~60 fps update rate
    window.limit_update_rate(Some(std::time::Duration::from_micros(16600)));

    let text:Vec<&str> = TEXT.split(",").collect();
    let mut bitmask = Vec::new();
    for _i in 0..6400 {
        bitmask.push(true);
        bitmask.push(false);
    }
    let cm = Mask::Color(0x00ff00);
    let bm = Mask::Bits(&bitmask);

    bitmaps[0].x = 0;
    bitmaps[0].y = 85;
    bitmaps[1].x = 0;
    bitmaps[1].y = 0;
    bitmaps[1].blit(&mut fb);
    draw_text(text[0], &mut fb, &mut bitmaps[0]);

    bitmaps[0].x = 160;
    bitmaps[0].y = 85;
    bitmaps[1].x = 240;
    bitmaps[1].y = 0;
    bitmaps[1].blit_mask(&mut fb, cm);
    draw_text(text[1], &mut fb, &mut bitmaps[0]);

    bitmaps[0].x = 160;
    bitmaps[0].y = 166;
    bitmaps[1].x = 240;
    bitmaps[1].y = 176;
    bitmaps[1].blit_mask(&mut fb, bm);
    draw_text(text[3], &mut fb, &mut bitmaps[0]);

    bitmaps[0].x = 85;
    bitmaps[0].y = 125;
    bitmaps[1].x = 35;
    bitmaps[1].y = 125;
    bitmaps[1].blit_part(&mut fb, 0, 40, 40);
    draw_text(text[4], &mut fb, &mut bitmaps[0]);

    bitmaps[0].x = 0;
    bitmaps[0].y = 196;
    bitmaps[1].x = -20;
    bitmaps[1].y = 206;
    bitmaps[1].blit(&mut fb);
    draw_text(text[2], &mut fb, &mut bitmaps[0]);

    while window.is_open() && !window.is_key_down(Key::Escape) {
        window
            .update_with_buffer(&fb.pixels, WIDTH, HEIGHT)
            .unwrap();
    }
}

// Text drawing
fn draw_text(text: &str, mut fb: &mut Framebuffer, font: &mut Bitmap) {
    let text = text.as_bytes();
    for i in 0..text.len() {
        font.blit_part(&mut fb, 8 * (text[i] as usize - 32), 8, 8);
        font.x += 8;
    }
}

examples/minifb.rs
//#![windows_subsystem = "windows"]
use minifb::{Key, Window, WindowOptions};
use blitter::*;

const WIDTH: usize = 640;
const HEIGHT: usize = 520;

fn main() {
    // Framebuffer initialization
    let mut pixels: Vec<u32> = vec!(0; WIDTH * HEIGHT);
    let mut fb = Framebuffer {width: WIDTH, height: HEIGHT, pixels: &mut pixels};
    
    // User bitmaps generating / loading
    let image:  Vec<u32> =  { vec![0xffffffff; 100] };
    let image2:  Vec<u32> =  { vec![0x0000ff00; 2500] };
    let image3:  Vec<u32> =  { vec![0x0000ff00; 100] };
    let path = "resources/test-image.png";
    let png = from_png_file(&path, PixelFormat::Zrgb).unwrap();

    // Test mask
    let test_mask = vec![true, false, true, false, true, false, true, false, true, false,
    false, true, false, true, false, true, false, true, false, true,
    true, false, true, false, true, false, true, false, true, false,
    false, true, false, true, false, true, false, true, false, true,
    true, false, true, false, true, false, true, false, true, false,
    false, true, false, true, false, true, false, true, false, true,
    true, false, true, false, true, false, true, false, true, false,
    false, true, false, true, false, true, false, true, false, true,
    true, false, true, false, true, false, true, false, true, false,
    false, true, false, true, false, true, false, true, false, true];
    let mask = Mask::Bits(&test_mask);

    // Bitmaps structs stored in a Vec (could be a hashmap or whatever you want), to give more easily ownership to other functions
    let mut bitmaps = Vec::new();
    bitmaps.push(Bitmap {w: 10, h: 10, x: 0, y: 0, pixels: &image});
    bitmaps.push(Bitmap{w: png.0, h: png.1, x: 0, y: 320, pixels: &png.2});
    bitmaps.push(Bitmap {w: 50, h: 50, x: 590, y: 470, pixels: &image2});
    bitmaps.push(Bitmap {w: 10, h: 10, x: 630, y: 460, pixels: &image3});

    let mut window = Window::new(
        "Test - ESC to exit",
        WIDTH,
        HEIGHT,
        WindowOptions::default(),
    )
    .unwrap_or_else(|e| {
        panic!("{}", e);
    });

    // Limit to max ~60 fps update rate
    window.limit_update_rate(Some(std::time::Duration::from_micros(16600)));

    // Drawing non moving part of screen
    bitmaps[1].blit(&mut fb);
    bitmaps[2].blit(&mut fb);
    fb.draw_fatpixel(WIDTH/2, HEIGHT/2, 4, 0xffffffff).unwrap();
    bitmaps[3].blit_mask(&mut fb, mask);

    while window.is_open() && !window.is_key_down(Key::Escape) {
        
        // We can easily give bitmaps and framebuffer ownership; of course you can it do the way you want
        move_square(&mut fb, &mut bitmaps);

        // We unwrap here as we want this code to exit if it fails. Real applications may want to handle this in a different way
        window
            .update_with_buffer(&fb.pixels, WIDTH, HEIGHT)
            .unwrap();
    }
}

// For testing : moves a 10x10 square and prints a 4x4 pixel at the center of the screen
fn move_square(mut fb: &mut Framebuffer, bitmaps: &mut Vec<Bitmap>) {
    // We just clear the animated part of the screen
    fb.clear_area(640, 10, 0, 0, 0).unwrap();

    bitmaps[0].blit(&mut fb);
    if bitmaps[0].x < WIDTH as isize - 10 { bitmaps[0].x = bitmaps[0].x+3; }
}


examples/square.rs
//#![windows_subsystem = "windows"]
use minifb::{Key, Window, WindowOptions};
use blitter::*;

const WIDTH: usize = 640;
const HEIGHT: usize = 480;

fn main() {
    // Framebuffer initialization
    let mut pixels: Vec<u32> = vec!(0; WIDTH * HEIGHT);
    let mut fb = Framebuffer {width: WIDTH, height: HEIGHT, pixels: &mut pixels};
    
    // Pixel data
    let image:  Vec<u32> =  { vec![0xffffffff; 100] };

    // Bitmap creation
    let mut bitmap = Bitmap {w: 10, h: 10, x: 0, y: 0, pixels: &image};
    
    let mut window = Window::new(
        "Test - ESC to exit",
        WIDTH,
        HEIGHT,
        WindowOptions::default(),
    )
    .unwrap_or_else(|e| {
        panic!("{}", e);
    });

    // Limit to max ~60 fps update rate
    window.limit_update_rate(Some(std::time::Duration::from_micros(16600)));

    while window.is_open() && !window.is_key_down(Key::Escape) {
        
        // We can easily give bitmaps and framebuffer ownership
        move_square(&mut fb, &mut bitmap);

        // We unwrap here as we want this code to exit if it fails. Real applications may want to handle this in a different way
        window
            .update_with_buffer(&fb.pixels, WIDTH, HEIGHT)
            .unwrap();
    }
}

// For testing : moves a 10x10 square and prints a 4x4 pixel at the center of the screen
fn move_square(mut fb: &mut Framebuffer, mut bitmap: &mut Bitmap) {
    // We just clear the animated part of the screen
    fb.clear_area(640, 10, 0, 0, 0).unwrap();

    bitmap.blit(&mut fb);
    if bitmap.x < WIDTH as isize - 10 { bitmap.x = bitmap.x+3; }
}


src/lib.rs
//! This library performs various blitting and drawing operations on a raw 32 bits framebuffer, whatever the encoding:
//! - Bitmap blitting / cropping
//! - Blit a part of bitmap (ie. bitmap fonts)
//! - Blit with a color or bits mask
//! - Pixel plotting
//! - Optional PNG decoding feature
//! 
//! Example:
//!```text
//! // Framebuffer initialization
//! let mut pixels: Vec<u32> = vec!(0; WIDTH * HEIGHT);
//! let mut fb = Framebuffer {width: WIDTH, height: HEIGHT, pixels: &mut pixels};
//!
//! // For example, you can push all the bitmaps in a single vec to give ownership of all bitmaps
//! let mut bitmaps = Vec::new();
//! bitmaps.push(Bitmap {w: 10, h: 10, x: 0, y: 0, pixels: &image::PIXELS});
//!
//! while *display loop with some display library* {
//!     blitter_test(&mut fb, &mut bitmaps);
//!     *your display lib display update function with buffer &fb.pixels*
//! }
//!
//! // For testing : moves a 10x10 square and prints a 4x4 pixel at the center of the screen
//! fn blitter_test(mut fb: &mut Framebuffer, bitmaps: &mut Vec<Bitmap>) {
//!     fb.clear_area(640, 10, 0, 0, 0).unwrap();
//!     bitmaps[0].blit(&mut fb);   //copies a bitmap to the framebuffer
//!     if bitmaps[0].x < WIDTH - 10 { bitmaps[0].x = bitmaps[0].x+3; } else { fb.clear(0); }
//!     fb.draw_fatpixel(320,240,4,0xffffffff).unwrap();
//! }
//! ```
//!
//! You can also view and run some (very basic) examples using the [minifb library](https://crates.io/crates/minifb) in the 'examples' directory:
//! ```text
//! cargo run --example demo --features="png-decode"
//! cargo run --example minifb --features="png-decode"
//! cargo run --example square
//! ```

use std::{fmt, result::Result};
#[cfg(feature = "png-decode")]
use {png::DecodingError, std::fs::File};

/// Output format of png decoding function
#[cfg(feature = "png-decode")]
pub enum PixelFormat {
    /// 0RGB
    Zrgb,
    /// RGBA, STRIP_ALPHA
    Rgba,
}

/// Mask applied to blitting operations
pub enum Mask<'a> {
    /// color mask (color acting as transparent in case of non alpha framebuffers)
    Color(u32),
    /// Bits mask (logical AND)
    Bits(&'a Vec<bool>),
    /// No mask
    None
}

/// To prevent buffer overflow
#[derive(Debug, PartialEq, Eq, Clone)]
pub enum BlitError {
    // Index out of bounds
    BlittingBeyondBoundaries,
}

impl fmt::Display for BlitError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("tzfile error: ")?;
        f.write_str(match self {
            BlitError::BlittingBeyondBoundaries => "You are blitting outside the framebuffer !",
        })
    }
}

/// This structure stores bitmap's sizes, coordinates, and a pointer to its pixel data
pub struct Bitmap<'a> {
    /// Bitmap width
    pub w: usize,
    /// Bitmap height
    pub h: usize,
    /// Bitmap horizontal position
    pub x: isize,
    /// Bitmap vertical position
    pub y: isize,
    /// 32 bits pixel data
    pub pixels: &'a Vec<u32>,
}

/// The framebuffer struct contains the buffer's width, height, and a pointer to its pixel data
pub struct Framebuffer<'a> {
    /// Framebuffer width
    pub width: usize,
    /// Framebuffer height
    pub height: usize,
    /// 32 bits pixel data
    pub pixels: &'a mut Vec<u32>,
}

struct ClippedCoords {
    // Non negative (unsigned) x coordinate
    ux: usize,
    // Non negative (unsigned) y coordinate
    uy: usize,
    // End of blit, x axis
    x_end: usize,
    // End of blit, y axis
    y_end: usize,
    // Pixel counter / start offset
    c: usize,
    // Pixel skip of source bitmap
    src_pixel_skip: usize,
}

impl Bitmap<'_> {
    /// Copies a bitmap to the framebuffer, applying a mask
    pub fn blit_mask(&self, fb: &mut Framebuffer, mask: Mask) {
        let mut cr = match self.compute_clipping(fb) {
            Some(c) => c,
            None => return
        };
        for inc_y in 0..cr.y_end {
            let x_offset: usize = inc_y * fb.width;
            let y_offset: usize = cr.uy * fb.width;
            for inc_x in 0..cr.x_end {
                match mask {
                    Mask::Color(c) => if self.pixels[cr.c] != c { fb.pixels[inc_x + x_offset + cr.ux + y_offset] = self.pixels[cr.c] },
                    Mask::Bits(b) => if b[cr.c] { fb.pixels[inc_x + x_offset + cr.ux + y_offset] = self.pixels[cr.c] },
                    Mask::None => fb.pixels[inc_x + x_offset + cr.ux + y_offset] = self.pixels[cr.c]
                }
                cr.c += 1;
            }
            cr.c += cr.src_pixel_skip;
        }
    }

    /// Convenience function to blit without any mask
    pub fn blit(&self, fb: &mut Framebuffer) {
        self.blit_mask(fb, Mask::None);
    }

    /// Copies a portion of a bitmap to the framebuffer
    pub fn blit_part(&self, fb: &mut Framebuffer, start_offset: usize, w: usize, h: usize) {
        let mut c = 0 + start_offset;
        // Temporary pixel buffer
        let mut t_pixels = vec![0; w * h];
        for inc_y in 0..h {
            for inc_x in 0..w {
                t_pixels[inc_x + inc_y * w ] = self.pixels[c];
                c += 1;
            }
            c += self.w - w;
        }
        // Temporary Bitmap; this way we can use the generic blit function
        let tx = self.x;
        let ty = self.y;
        let tw = w;
        let th = h;
        let t = Bitmap { x: tx, y: ty, w: tw, h: th, pixels: &t_pixels};
        t.blit(fb);
    }

    fn compute_clipping(&self, fb: &Framebuffer) -> Option<ClippedCoords> {
        // Are x or y negative values ? compute cropped pixels size and convert x and y to unsigned values
        let ux = if self.x > 0 { self.x as usize } else { 0 };
        let uy = if self.y > 0 { self.y as usize } else { 0 };
        let cropped_x = self.x.abs() as usize;
        let cropped_y = self.y.abs() as usize;
        // Need to crop the top of the bitmap
        let r = if ux + self.w <= fb.width && uy + self.h < fb.height && self.x >= 0 && self.y < 0 && (self.y + self.h as isize) > 0 {
            //println!("Cropping top");
            Some(ClippedCoords {
                x_end: self.w,
                y_end: self.h - cropped_y,
                src_pixel_skip: 0,
                c: cropped_y * self.w,
                ux: ux,
                uy: uy,
            })
        }
        // Need to crop the top left of the bitmap
        else if self.x < 0 && self.y < 0 && self.x + self.w as isize > 0 && self.y + self.h as isize > 0 {
            //println!("Cropping top left");
            Some(ClippedCoords {
                x_end: self.w - cropped_x,
                y_end: self.h - cropped_y,
                src_pixel_skip: cropped_x,
                c: cropped_y * self.w + cropped_x,
                ux: ux,
                uy: uy,
            })
        }
        // Need to crop the top right of the bitmap
        else if ux + self.w > fb.width && ux < fb.width && self.y < 0 && self.y + self.h as isize > 0 {
            //println!("Cropping top right");
            Some(ClippedCoords {
                x_end: fb.width - ux,
                y_end: self.h - cropped_y,
                src_pixel_skip: self.w - (fb.width - ux),
                c: cropped_y * self.w,
                ux: ux,
                uy: uy,
            })
        }
        // Need to crop the bottom left of the bitmap
        else if uy + self.h > fb.height && self.x < 0 && self.x + self.w as isize > 0 && uy < fb.height {
            //println!("Cropping bottom left");
            Some(ClippedCoords {
                x_end: self.w - cropped_x,
                y_end: fb.height - uy,
                src_pixel_skip: cropped_x,
                c: cropped_x,
                ux: ux,
                uy: uy,
            })
        }
        // Need to crop the bottom right of the bitmap
        else if ux + self.w > fb.width && uy + self.h > fb.height && ux < fb.width && uy < fb.height {
            //println!("Cropping bottom right");
            Some(ClippedCoords {
                x_end: fb.width - ux,
                y_end: fb.height - uy,
                src_pixel_skip: self.w - (fb.width - ux),
                c: 0,
                ux: ux,
                uy: uy,
            })
        }
        // Need to crop the bottom of the bitmap
        else if ux + self.w < fb.width && self.x + self.w as isize > 0 && uy + self.h > fb.height && uy < fb.height {
            //println!("Cropping bottom");
            Some(ClippedCoords {
                x_end: self.w,
                y_end: fb.height - uy,
                src_pixel_skip: 0,
                c: 0,
                ux: ux,
                uy: uy,
            })
        }
        // Need to crop the left of the bitmap
        else if self.x < 0 && self.x + self.w as isize > 0 && self.y > 0 && uy < fb.height {
            //println!("Cropping left");
            Some(ClippedCoords {
                x_end: self.w - cropped_x,
                y_end: self.h,
                src_pixel_skip: cropped_x,
                c: cropped_x,
                ux: ux,
                uy: uy,
            })
        }
        // Need to crop the right of the bitmap
        else if ux + self.w > fb.width && self.y >= 0 && ux <= fb.width && uy + self.h < fb.height {
            //println!("Cropping right");
            Some(ClippedCoords {
                x_end: fb.width - ux,
                y_end: self.h,
                src_pixel_skip: self.w - (fb.width - ux),
                c: 0,
                ux: ux,
                uy: uy,
            })
        }
        // Blitting outside the screen -> no need to blit anything
        else if ux > fb.width || uy > fb.height || (self.x + self.w as isize) < 0 || (self.y + self.h as isize) < 0 {
            //println!("Outside framebuffer");
            None
        }
        // No need to crop      self.x + self.w <= fb.width && self.y + self.h <= fb.height
        else {
            //println!("No cropping");
            Some(ClippedCoords {
                x_end: self.w,
                y_end: self.h,
                src_pixel_skip: 0,
                c: 0,
                ux: ux,
                uy: uy,
            })
        };
        r
    }
}

impl Framebuffer<'_> {
    /// Partial clear of the framebuffer
    pub fn clear_area(
        &mut self,
        w: usize,
        h: usize,
        x: usize,
        y: usize,
        clear_color: u32,
    ) -> Result<(), BlitError> {
        if ((x + w) * (y + h)) > self.pixels.len() {
            return Err(BlitError::BlittingBeyondBoundaries);
        };
        for inc_y in 0..h {
            let x_offset: usize = inc_y * self.width;
            let y_offset: usize = y * self.width;
            for inc_x in 0..w {
                self.pixels[inc_x + x_offset + x + y_offset] = clear_color;
            }
        }
        Ok(())
    }

    /// Complete clear of the framebuffer
    pub fn clear(&mut self, clear_color: u32) {
        for inc_x in 0..self.width * self.height {
            self.pixels[inc_x] = clear_color;
        }
    }

    /// Drawing a pixel
    pub fn draw_pixel(&mut self, x: usize, y: usize, color: u32) -> Result<(), BlitError> {
        if x > self.width || y > self.height {
            return Err(BlitError::BlittingBeyondBoundaries);
        };
        self.pixels[x + y * self.width] = color;
        Ok(())
    }

    /// Drawing a fat pixel
    pub fn draw_fatpixel(
        &mut self,
        x: usize,
        y: usize,
        size: usize,
        color: u32,
    ) -> Result<(), BlitError> {
        if x > self.width - size || y > self.height - size {
            return Err(BlitError::BlittingBeyondBoundaries);
        };
        self.clear_area(size, size, x, y, color)?;
        Ok(())
    }
}

#[cfg(feature = "png-decode")]
/// Creates a tuple containing width, height, and pixel data from a PNG file
pub fn from_png_file(
    pngfile: &str,
    pxfmt: PixelFormat,
) -> Result<(usize, usize, Vec<u32>), DecodingError> {
    let shift: u32 = match pxfmt {
        PixelFormat::Zrgb => 0,
        PixelFormat::Rgba => 8,
    };
    // The default output transformation is `Transformations::EXPAND | Transformations::STRIP_ALPHA`.
    let decoder = png::Decoder::new(File::open(&pngfile)?);
    let (info, mut reader) = decoder.read_info()?;
    let dpth = match info.color_type {
        png::ColorType::RGB => 3,
        png::ColorType::RGBA => 4,
        _ => return Err(DecodingError::Format(std::borrow::Cow::Borrowed("Function only handles RGB and RGBA formats")))
    };
    // Allocate the output buffer.
    let mut buf = vec![0; info.buffer_size()];
    // Read the next frame. Currently this function should only called once.
    reader.next_frame(&mut buf)?;
    // convert buffer to u32
    let u32_buffer: Vec<u32> = buf
        .chunks(dpth)
        .map(|v| ((v[0] as u32) << 16) | ((v[1] as u32) << 8) | v[2] as u32)
        .map(|x| x << shift)
        .collect();

    Ok((info.width as usize, info.height as usize, u32_buffer))
}

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
        assert_eq!(2 + 2, 4);
    }
}
